textscreen/txt_sdl.c

```diff
@@ -0,0 +1,1009 @@
+    SDL_UnlockSurface(screenbuffer);
+
+    screendata = malloc(TXT_SCREEN_W * TXT_SCREEN_H * 2);
+    memset(screendata, 0, TXT_SCREEN_W * TXT_SCREEN_H * 2);
+
+    return 1;
+}
+
+void TXT_Shutdown(void)
+{
+    free(screendata);
+    screendata = NULL;
+    SDL_FreeSurface(screenbuffer);
+    screenbuffer = NULL;
+    SDL_QuitSubSystem(SDL_INIT_VIDEO);
+}
+
+void TXT_SetColor(txt_color_t color, int r, int g, int b)
+{
+    SDL_Color c = {r, g, b, 0xff};
+
+    SDL_LockSurface(screenbuffer);
+    SDL_SetPaletteColors(screenbuffer->format->palette, &c, color, 1);
+    SDL_UnlockSurface(screenbuffer);
+}
+
+unsigned char *TXT_GetScreenData(void)
+{
+    return screendata;
+}
+
+static inline void UpdateCharacter(int x, int y)
+{
+    unsigned char character;
+    const uint8_t *p;
+    unsigned char *s, *s1;
+    unsigned int bit;
+    int bg, fg;
+    unsigned int x1, y1;
+
+    p = &screendata[(y * TXT_SCREEN_W + x) * 2];
+    character = p[0];
+
+    fg = p[1] & 0xf;
+    bg = (p[1] >> 4) & 0xf;
+
+    if (bg & 0x8)
+    {
+        // blinking
+
+        bg &= ~0x8;
+
+        if (((SDL_GetTicks() / BLINK_PERIOD) % 2) == 0)
+        {
+            fg = bg;
+        }
+    }
+
+    // How many bytes per line?
+    p = &font->data[(character * font->w * font->h) / 8];
+    bit = 0;
+
+    s = ((unsigned char *) screenbuffer->pixels)
+      + (y * font->h * screenbuffer->pitch)
+      + (x * font->w);
+
+    for (y1=0; y1<font->h; ++y1)
+    {
+        s1 = s;
+
+        for (x1=0; x1<font->w; ++x1)
+        {
+            if (*p & (1 << bit))
+            {
+                *s1++ = fg;
+            }
+            else
+            {
+                *s1++ = bg;
+            }
+
+            ++bit;
+            if (bit == 8)
+            {
+                ++p;
+                bit = 0;
+            }
+        }
+
+        s += screenbuffer->pitch;
+    }
+}
+
+static int LimitToRange(int val, int min, int max)
+{
+    if (val < min)
+    {
+        return min;
+    }
+    else if (val > max)
+    {
+        return max;
+    }
+    else
+    {
+        return val;
+    }
+}
+
+static void GetDestRect(SDL_Rect *rect)
+{
+    int w, h;
+
+    SDL_GetRendererOutputSize(renderer, &w, &h);
+    rect->x = 0;
+    rect->y = 0;
+    rect->w = screenbuffer->w;
+    rect->h = screenbuffer->h;
+}
+
+void TXT_UpdateScreenArea(int x, int y, int w, int h)
+{
+    SDL_Texture *screentx;
+    SDL_Rect rect;
+    int x1, y1;
+    int x_end;
+    int y_end;
+
+    SDL_LockSurface(screenbuffer);
+
+    x_end = LimitToRange(x + w, 0, TXT_SCREEN_W);
+    y_end = LimitToRange(y + h, 0, TXT_SCREEN_H);
+    x = LimitToRange(x, 0, TXT_SCREEN_W);
+    y = LimitToRange(y, 0, TXT_SCREEN_H);
+
+    for (y1=y; y1<y_end; ++y1)
+    {
+        for (x1=x; x1<x_end; ++x1)
+        {
+            UpdateCharacter(x1, y1);
+        }
+    }
+
+    SDL_UnlockSurface(screenbuffer);
+
+    SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "linear");
```

Comment:
I guess for scaling up of blocky DOS font and pixel art, "nearest" makes more sense here.

Reply:
I tried it and it looks bad. Technically we have the wrong aspect ratio, but scaling the font is tricky. I think 1:1 linear scaling works best.

Reply:
Okay, fine with me then.